velox/exec/PartitionedOutputBufferManager.cpp

```diff
@@ -275,12 +371,83 @@ BlockingReason PartitionedOutputBuffer::enqueue(
                  : BlockingReason::kNotBlocked;
 }
 
+void PartitionedOutputBuffer::enqueueBroadcastOutputLocked(
+    std::unique_ptr<SerializedPage> data,
+    std::vector<DataAvailable>& dataAvailableCbs) {
+  VELOX_DCHECK(isBroadcast());
+  VELOX_CHECK_NULL(arbitraryBuffer_);
+  VELOX_DCHECK(dataAvailableCbs.empty());
+
+  std::shared_ptr<SerializedPage> sharedData(data.release());
+  for (auto& buffer : buffers_) {
+    if (buffer != nullptr) {
+      buffer->enqueue(sharedData);
+      dataAvailableCbs.emplace_back(buffer->getAndClearNotify());
+    }
+  }
+
+  // NOTE: we don't need to add new buffer to 'dataToBroadcast_' if there is no
+  // more output buffers.
+  if (!noMoreBuffers_) {
+    dataToBroadcast_.emplace_back(sharedData);
+  }
+}
+
+void PartitionedOutputBuffer::enqueueArbitraryOutputLocked(
+    std::unique_ptr<SerializedPage> data,
+    std::vector<DataAvailable>& dataAvailableCbs) {
+  VELOX_DCHECK(isArbitrary());
+  VELOX_DCHECK_NOT_NULL(arbitraryBuffer_);
+  VELOX_DCHECK(dataAvailableCbs.empty());
+  VELOX_CHECK(!arbitraryBuffer_->hasNoMoreData());
+
+  arbitraryBuffer_->enqueue(std::move(data));
+  VELOX_CHECK_LT(nextArbitraryLoadBufferIndex_, buffers_.size());
+  int32_t bufferId = nextArbitraryLoadBufferIndex_;
+  for (int32_t i = 0; i < buffers_.size();
+       ++i, bufferId = (bufferId + 1) % buffers_.size()) {
+    if (arbitraryBuffer_->empty()) {
+      nextArbitraryLoadBufferIndex_ = bufferId;
+      break;
+    }
+    auto* buffer = buffers_[bufferId].get();
+    if (buffer == nullptr) {
+      continue;
+    }
+    buffer->maybeLoadData(arbitraryBuffer_.get());
+    dataAvailableCbs.emplace_back(buffer->getAndClearNotify());
+  }
+}
```

Comment:
So we copy the data to `ArbitraryBuffer`, then distribute to one of the `DestinationBuffer`. Can we completely get rid of the middle layer of `ArbitraryBuffer`

Reply:
There is no data copy between arbitrary buffer and destination buffer and we just move the smarter pointers around and the cost is trivial. ArbitraryBuffer buffer provides data load abstraction which simplifies the implementation. The data load logic is more complicated than broadcast case so it is good to have a soft abstraction on top of a queue.